### PR TITLE
CumulativeToDelta conversion support for metricpipelines

### PR DESCRIPTION
**Pull request by Teneroy**
**Description**

Changes proposed in this pull request:

- Added cumulativeToDelta processor
- Introduced new flag in metricpipeline CRD for cumulative to delta conversion
- Added new Unit and e2e tests to test cumulative to delta conversion

**Related issue(s)**

#207

### CODE REVIEW

File: config/crd/bases/telemetry.kyma-project.io_metricpipelines.yaml [hunk 1]
```diff
@@ -45,6 +45,10 @@ spec:
               output:
                 description: Configures the trace receiver of a MetricPipeline.
                 properties:
+                  cumulative:
```
Comment (chrkl) [inline config/crd/bases/telemetry.kyma-project.io_metricpipelines.yaml:48]: Improve naming

File: internal/otelcollector/config/config.go [hunk 2]
```diff
@@ -59,6 +59,16 @@ type BatchProcessorConfig struct {
 	SendBatchMaxSize int    `yaml:"send_batch_max_size"`
 }
 
+type CumulativeToDeltaConfig struct {
```
Comment (chrkl) [inline internal/otelcollector/config/config.go:62]: Remove Include and Exclude

File: internal/reconciler/metricpipeline/config_builder_test.go [hunk 3]
```diff
@@ -296,10 +303,11 @@ service:
             receivers:
                 - otlp
             processors:
-                - memory_limiter
+                - batch
+                - cumulativetodelta/test
```
Comment (chrkl) [inline internal/reconciler/metricpipeline/config_builder_test.go:307]: Remove /test suffix

File: internal/resources/common/resources.go [hunk 4]
```diff
@@ -32,3 +32,25 @@ func MakeClusterRoleBinding(name types.NamespacedName) *v1.ClusterRoleBinding {
 	}
 	return &clusterRoleBinding
 }
+
+func MakeClusterRole(name types.NamespacedName) *v1.ClusterRole {
```
Comment (chrkl) [inline internal/resources/common/resources.go:36]: Already included in main

File: internal/reconciler/metricpipeline/config_builder.go [hunk 5]
```diff
@@ -59,6 +64,13 @@ func makeOtelCollectorConfig(ctx context.Context, c client.Reader, pipelines []v
 	}, allVars, nil
 }
 
+func getProcessorAliases(output *v1alpha1.MetricPipelineOutput) []string {
+	if output.ToDelta {
+		return []string{"cumulativetodelta", "memory_limiter", "k8sattributes", "resource", "batch"}
```
Comment (chrkl) [inline internal/reconciler/metricpipeline/config_builder.go:69]: Do some investigation if the order of the processors matter. Iirc, the `memory_limiter` should be the first processor.

File: apis/telemetry/v1alpha1/metricpipeline_types.go [hunk 6]
```diff
@@ -89,6 +89,8 @@ type MetricPipelineIstioInput struct {
 type MetricPipelineOutput struct {
 	// Defines an output using the OpenTelemetry protocol.
 	Otlp *OtlpOutput `json:"otlp"`
+	// Defines whether this metricpipeline should convert monotonic, cumulative sum and histogram metrics to monotonic, delta metrics.
```
Comment (NHingerl) [inline apis/telemetry/v1alpha1/metricpipeline_types.go:92]: [Camel case for Kubernetes resources](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/02-style-and-terminology.md).
```suggestion
	// Defines whether this MetricPipeline should convert monotonic, cumulative sum and histogram metrics to monotonic, delta metrics.
```

File: config/crd/bases/telemetry.kyma-project.io_metricpipelines.yaml [hunk 7]
```diff
@@ -194,6 +194,11 @@ spec:
                     required:
                     - endpoint
                     type: object
+                  toDelta:
+                    description: Defines whether this metricpipeline should convert
```
Comment (NHingerl) [inline config/crd/bases/telemetry.kyma-project.io_metricpipelines.yaml:198]:
```suggestion
                    description: Defines whether this MetricPipeline should convert
```

File: docs/user/resources/05-metricpipeline.md [hunk 8]
```diff
@@ -85,6 +85,7 @@ For details, see the [MetricPipeline specification file](https://github.com/kyma
 | **output.&#x200b;otlp.&#x200b;headers.&#x200b;valueFrom.&#x200b;secretKeyRef.&#x200b;name**  | string |  |
 | **output.&#x200b;otlp.&#x200b;headers.&#x200b;valueFrom.&#x200b;secretKeyRef.&#x200b;namespace**  | string |  |
 | **output.&#x200b;otlp.&#x200b;protocol**  | string | Defines the OTLP protocol (http or grpc). Default is GRPC. |
+| **output.&#x200b;toDelta**  | boolean | Defines whether this metricpipeline should convert monotonic, cumulative sum and histogram metrics to monotonic, delta metrics. |
```
Comment (NHingerl) [inline docs/user/resources/05-metricpipeline.md:88]:
```suggestion
| **output.&#x200b;toDelta**  | boolean | Defines whether this MetricPipeline should convert monotonic, cumulative sum and histogram metrics to monotonic, delta metrics. |
```

File: test/e2e/metrics_test.go [hunk 9]
```diff
@@ -134,6 +134,73 @@ var _ = Describe("Metrics", func() {
 
 	})
 
+	Context("When a metricpipeline has toDelta flag active", Ordered, func() {
```
Comment (NHingerl) [inline test/e2e/metrics_test.go:137]:
```suggestion
	Context("When a MetricPipeline has toDelta flag active", Ordered, func() {
```

**Review by NHingerl — COMMENTED** [review]
Thinking about it; I agree with the change to convertToDelta. It's describing more transparently what the flag does.

**Review by NHingerl — DISMISSED** [review]

**Review by NHingerl — DISMISSED** [review]

File: internal/otelcollector/config/builder/metric/gateway/config_builder.go [hunk 10]
```diff
@@ -101,3 +104,10 @@ func makeServiceConfig(pipelines config.PipelinesConfig) config.ServiceConfig {
 		Extensions: []string{"health_check", "pprof"},
 	}
 }
+
+func getProcessorAliases(output *v1alpha1.MetricPipelineOutput) []string {
+	if output.ConvertToDelta {
+		return []string{"memory_limiter", "k8sattributes", "resource", "batch", "cumulativetodelta"}
```
Comment (chrkl) [inline internal/otelcollector/config/builder/metric/gateway/config_builder.go:110]: Keep the batch processor at the end of the list

File: test/e2e/testkit/matchers/metric_matchers.go [hunk 11]
```diff
@@ -34,6 +34,36 @@ func HaveMetrics(expectedMetrics ...pmetric.Metric) types.GomegaMatcher {
 	}, gomega.ContainElements(expectedMetrics))
 }
 
+func HaveSumMetrics(expectedMetrics ...pmetric.Metric) types.GomegaMatcher {
```
Comment (chrkl) [inline test/e2e/testkit/matchers/metric_matchers.go:37]: Add unittests for the. Take a look how the other matchers are tested in metric_matchers_test.go.

**Review by chrkl — APPROVED** [review]

**Review by nataliasitko — APPROVED** [review]

**Pull request merged**